### ISSUE TITLE
Recommendations: log outcomes and add lightweight single-user learned ranking

### ISSUE THREAD

**Issue by evchuk4018** [evchuk4018/navidrome#12]
## Goal

Add a lightweight **single-user learned ranking layer** that learns from the user's own recommendation outcomes instead of permanently hand-tuning every coefficient in `recommendations.DefaultWeights()`.

This is not a neural-network ticket.

The first model should be small, inspectable, CPU-cheap, restart-safe, and guarded against overfitting.

## Current repo state

The fork already has a strong feature foundation:

`core/recommendations/ranking.go` currently exposes:
- provider similarity
- seed affinity
- session affinity
- play history
- recent listening
- starred
- recency
- fatigue
- contextual transition affinity

Personal Radio also has:
- explicit `started`, `threshold_reached`, `completed`, and skip feedback
- persisted `PersonalRadioItem.PlaybackOutcome`
- contextual transition aggregates
- traceable score breakdowns

The missing piece is an exposure/outcome dataset linking:

> “these were the features when candidate X was selected”

to:

> “the user accepted/completed/skipped X”

Without that dataset, weights such as `TransitionAffinity: 1.5` can only be tuned manually.

## Single-user constraints

Do not build distributed ML infrastructure.

Do not use cross-user data.

Do not require a GPU or external model server.

The model must work with hundreds-to-thousands of examples, not millions.

## Phase 1: persist recommendation impressions

Add a migration with a table similar to:

```sql
create table recommendation_impression (
    id                     varchar(255) primary key,
    user_id                varchar(255) not null references user(id) on delete cascade,
    surface                varchar(32) not null,
    surface_item_id        varchar(255) not null default '',
    session_id             varchar(255) not null default '',
    seed_key               varchar(512) not null default '',
    candidate_key          varchar(512) not null,

    provider_similarity    real not null default 0,
    taste_affinity         real not null default 0,
    seed_affinity          real not null default 0,
    session_affinity       real not null default 0,
    transition_affinity    real not null default 0,
    familiarity            real not null default 0,
    fatigue                real not null default 0,
    recency                real not null default 0,
    starred                real not null default 0,

    base_score             real not null,
    selected_position      integer not null default -1,

    outcome                varchar(32) not null default '',
    listened_ms            integer not null default 0,
    duration_ms            integer not null default 0,

    created_at             datetime not null,
    outcome_at             datetime
);

create index recommendation_impression_training
    on recommendation_impression(user_id, outcome, created_at);
create index recommendation_impression_surface_item
    on recommendation_impression(user_id, surface, surface_item_id);
```

If #10/#13 are not implemented yet, their feature columns can remain zero until available.

Prefer explicit feature columns over an opaque JSON blob for the initial model so migrations/tests/debugging remain straightforward.

### Personal Radio

Record an impression when a candidate is actually selected by `composeRadioCandidates` and materialized into a `PersonalRadioItem`, not merely when it appeared in the 200-candidate retrieval pool.

Store `surface='personal_radio'` and `surface_item_id = personal_radio_item.id`.

When `Service.Feedback()` finalizes/strengthens the playback outcome, update the matching impression:

- accepted
- completed
- early_skip
- late_skip
- keep

Use the already-idempotent radio playback outcome semantics.

### Quick Pick

Initial implementation may record a Quick Pick impression when an item is returned by the API, but outcome should only be attached after actual interaction.

Add a stable `impressionId` to the Quick Pick item payload and report the click/play from `ui/src/quickpick/QuickPick.jsx`.

Do not treat “returned in API response” as a positive outcome.

## Training target

Start with a simple binary target:

Positive:
- accepted
- completed
- keep

Negative:
- early_skip

Late skip should initially be excluded from model training or treated as weak/neutral, not a hard negative.

Completed may optionally receive higher sample weight than threshold-only accepted.

## Model

Implement regularized logistic regression in Go.

Features should be normalized/bounded inputs already produced by the ranker.

Conceptually:

```text
p_accept = sigmoid(
    b
  + w_similarity   * provider_similarity
  + w_taste        * taste_affinity
  + w_transition   * transition_affinity
  + w_session      * session_affinity
  + w_seed         * seed_affinity
  + w_familiarity  * familiarity
  - w_fatigue      * fatigue
  + ...
)
```

The implementation can use small-batch gradient descent or another deterministic lightweight optimizer.

Do not introduce Python/TensorFlow/PyTorch runtime dependencies.

## Safety against overfitting

Mandatory guards:

- do not activate learned ranking below a minimum labeled sample count
- suggested initial threshold: 200 labeled impressions
- use L2 regularization
- clamp coefficients to a conservative range
- train on chronological history so future examples are not used to predict the past
- hold out the newest ~20% of examples for evaluation
- only activate a newly trained model if it is at least as good as the current active model on held-out log loss
- persist the previous active model for rollback

Because this is one user's data, prefer stability over aggressive adaptation.

## Blend, do not instantly replace the handcrafted ranker

Initial production scoring should blend:

```text
final_score =
    (1 - alpha) * normalized_base_rank
  + alpha       * learned_accept_score
```

Increase `alpha` with available labeled data, for example:

- <200 labels: 0
- 200–500: up to .20
- 500–1500: up to .35
- >1500: cap around .50 initially

The exact schedule is implementation-tunable, but the model must not suddenly replace all handcrafted logic after the first training run.

Transition suppression, hard duplicate exclusion, and radio composition constraints remain outside the ML model.

## Model persistence

Add a table such as:

```sql
create table recommendation_model (
    user_id          varchar(255) not null references user(id) on delete cascade,
    model_version    integer not null,
    coefficients_json text not null,
    intercept        real not null,
    training_rows    integer not null,
    validation_loss  real not null,
    trained_at       datetime not null,
    active           integer not null default 0,
    primary key(user_id, model_version)
);
```

JSON is acceptable for model coefficients because this is model state rather than queryable event data.

## Provider reliability

Use the same impression dataset to compute per-provider single-user success statistics later.

If issue #11 exposes `Options.ProviderReliability`, this ticket may derive conservative reliability values from impressions only after enough observations per provider.

Do not make provider reliability a blocker for the first learned model.

## Training cadence

For a single-user app:

- train after startup if enough new labeled rows accumulated
- retrain after e.g. 50 additional labels or once per day
- no more than one training job at a time
- training should complete quickly on CPU
- failure leaves the previous model active

## Explainability

For any ranked/selected candidate, trace:

- base handcrafted score
- learned probability
- blend alpha
- final blended score
- active model version

Expose learned coefficients in debug logs or an internal diagnostic method.

### Impression lifecycle
- selected Personal Radio candidate creates exactly one impression
- duplicate refill/planning does not duplicate the same materialized-item impression
- radio feedback updates the matching impression idempotently
- completed outcome cannot later be downgraded by duplicate skip callback
- Quick Pick outcome only records after interaction

### Training
Synthetic data where transition affinity predicts acceptance should produce a positive transition coefficient.

Synthetic data where fatigue predicts skips should produce a negative effective relationship.

- coefficient bounds respected
- L2 regularization applied
- chronological train/validation split deterministic
- model not activated below minimum data
- worse validation model does not replace active model

### Integration
- alpha=0 reproduces current handcrafted ordering
- learned score can alter ordering after sufficient data
- hard suppression/duplicate rules still apply
- radio Familiar/Balanced/Discover composition remains post-ranking logic

### Persistence
- active model survives restart
- training resumes from persisted impressions
- failed training keeps prior active model

## Out of scope

- neural networks / transformers
- embeddings
- collaborative filtering
- cross-user training
- reinforcement learning
- automatically changing radio mode
- removing score breakdowns/manual weights

## Acceptance criteria

- [ ] Selected recommendations are persisted with the feature values used at selection time.
- [ ] Radio outcomes are linked back to impressions idempotently.
- [ ] A lightweight regularized single-user model can be trained locally.
- [ ] Learned ranking stays disabled below a safe minimum dataset size.
- [ ] Learned score is blended gradually with the existing ranker.
- [ ] Validation prevents obviously worse models from replacing the active model.
- [ ] Model state survives restart.
- [ ] No external ML runtime/service is required.
- [ ] Existing hard filters, transition semantics, and queue composition remain authoritative.
